### PR TITLE
Fix heap walking with global allocation context

### PR DESCRIPTION
**Pull request by cshung**
In [case](https://github.com/dotnet/runtime/blob/main/src/coreclr/vm/gcheaputilities.h#L127) we are on Windows, x86 or amd64, single-core, we will use the global allocation context instead of the thread allocation context for allocating memory. The debugger needs to be aware of that when it walks the heap and skips the appropriate portions of it.

This change DACize the `g_global_alloc_context` and make sure the `mAllocInfo` [here](https://github.com/dotnet/runtime/blob/26372410d43ebb19220cce868a6f87186f88b2ce/src/coreclr/debug/daccess/dacdbiimpl.cpp#L6527) is populated with the global allocation context as well. With that, [CheckAllocAndSegmentRange](https://github.com/dotnet/runtime/blob/26372410d43ebb19220cce868a6f87186f88b2ce/src/coreclr/debug/daccess/dacdbiimpl.cpp#L6503) would properly skip the global allocation context when it reaches it and do the right thing.

The global allocation context information is also exposed through a new interface ISOSDacInterface12 so that it can be consumed by SOS [here](https://github.com/dotnet/diagnostics/pull/2949).

### CODE REVIEW

File: src/coreclr/inc/clrconfigvalues.h [hunk 1]
```diff
@@ -280,6 +280,13 @@ RETAIL_CONFIG_DWORD_INFO(UNSUPPORTED_HeapVerify, W("HeapVerify"), 0, "When set v
 RETAIL_CONFIG_DWORD_INFO(UNSUPPORTED_GCNumaAware, W("GCNumaAware"), 1, "Specifies if to enable GC NUMA aware")
 RETAIL_CONFIG_DWORD_INFO(EXTERNAL_GCCpuGroup, W("GCCpuGroup"), 0, "Specifies if to enable GC to support CPU groups")
 RETAIL_CONFIG_STRING_INFO(EXTERNAL_GCName, W("GCName"), "")
+/**
+ * This flag allows us to force the runtime to use global allocation context on Windows x86/amd64 instead of thread allocation context just for testing purpose.
+ * The flag is unsafe for a subtle reason. Although the access to the g_global_alloc_context is protected under a lock. The implementation of
+ * that lock in the JIT helpers are not multi-core safe (in particular, it used and inc instruction without using the LOCK prefix). This is
+ * only useful for ad-hoc testing.
+ */
+CONFIG_DWORD_INFO(INTERNAL_GCUseGlobalAllocationContext, W("GCUseGlobalAllocationContext"), 0, "Force using the global allocation context for testing only")
```
Comment (hoyosjs) [inline src/coreclr/inc/clrconfigvalues.h:289]: Is using `DOTNET_PROCESSOR_COUNT` not sufficient to give you the behavior?

Comment (cshung) [reply src/coreclr/inc/clrconfigvalues.h:289]: Nope, it doesn't change `g_SystemInfo.dwProcessorCount` and won't impact the behavior at all.

Comment (hoyosjs) [reply src/coreclr/inc/clrconfigvalues.h:289]: Separately, if that defines the number of compute cores available, should we be using the global alloc context in that scneario?

Comment (cshung) [reply src/coreclr/inc/clrconfigvalues.h:289]: As the comment indicates, the implementation of global allocation context is not multi-core safe, and it is by design for performance. Unless that setting will restrict the process from actually using more than one core. This is not safe.

Comment (hoyosjs) [reply src/coreclr/inc/clrconfigvalues.h:289]: That setting is meant to restrict how many cores a dotnet process can use, so it should be safe

**Comment by clguiman** [comment]
Are there plans to use the new DAC API in the DBI heap enumeration APIs? (ICorDebugHeapEnum::Next)?

**Review by hoyosjs — APPROVED** [review]

**Pull request merged**